### ISSUE TITLE
Proposal: New DockManager Control

### ISSUE THREAD

**Issue by shaggygi** [microsoft/microsoft-ui-xaml#668]
## Summary
It would be nice to have a new DockManager control similar to how Visual Studio and [AvalonDock](https://github.com/xceedsoftware/wpftoolkit/wiki/AvalonDock) works.

## Rationale
This type of control can be used for many LOB-type apps and has been missing as a 1st-class control for some time.  With my limited knowledge of where WinUI is headed, it does seem like a control that would be helpful.

**Comment by mdtauk** [comment]
Would this control need a DockablePanel control which contains XAML that fills the space, and then that panel handles the dragging and triggers the hint icons on the DockManager panel which contains the app's UI?

**Comment by shaggygi** [comment]
@mdtauk Great point.  I think there will need to be a few supporting controls, as well.  Similar to TabItem under the TabControl.

**Comment by jevansaks** [comment]
Bringing @harvinders proposal's description over here:

# Proposal: Dockable Windows 


## Summary
Add support for dockable windows that can be docked to a window like what we have in Visual Studio 

## Rationale

* This is one of a popular Desktop apps modes, which UWP is currently lacking
* Allows us to make apps who's layout is personalised by the user





## Scope

| Capability | Priority |
| :---------- | :------- |
| Dock window  | Must |
| Dock window container control, inside which the dock window docks | Must |
| Dock window container (inside) dock options (top,bottom,center,left,right) | Must |
| Undock options (float) | Must |
| Dock options to any window (outside) (top,bottom,left,right) | Should |

**Comment by YuliKl** [comment]
Related feature request - #1080 

We definitely need additional innovation in this space but won't be able to begin implementation until after WinUI 3.

**Comment by shaggygi** [comment]
@YuliKl 

> We definitely need additional innovation in this space but won't be able to begin implementation until after WinUI 3.

That is completely understandable.  I'm very glad the team is considering this control.  Look forward to WinUI 3.0 and future goodness 😄

**Comment by carmineos** [comment]
Both WinForms ([dockpanelsuite](https://github.com/dockpanelsuite/dockpanelsuite)) and WPF ([AvalonDock](https://github.com/Dirkster99/AvalonDock)) have widely used third-party controls for docking. Another good example would be [UWP Docking Control by Syncfusion](https://www.syncfusion.com/uwp-ui-controls/docking). Having a control like this would make easier to migrate apps to WinUI.

**Comment by LouChiShu** [comment]
Is there any progress on this feature? It seems like there hasn't been update in over a year and I would say that this seems like a pretty critical feature is Microsoft are truly aiming to make WinUI 3 the new default UI for Windows apps. Right now I am in the early stages of writing some graphical profiling software and evaluating what software is available out there, and not having an easy method of moving panels/dockable windows between areas to match the users needs seems like a huge oversight. I asked about this feature in #4784 and was told to put forward a feature request but it seems like the request is already here, it's just not made any progress.

They way I see it is that if I wanted to have what most would consider a pretty standard feature in software tools I would be best off using some combination of SDL2/Dear Imgui or a third party library for WPF/WinForms. This means that I wouldn't be able to take advantage of other WinUI 3 features which is a shame as for the most part WinUI 3 seems like a solid UI framework.

Is this feature still in consideration, or would it be something that is likely to be left to third parties again as with dockpanelsuite and AvalonDock?

**Comment by znakeeye** [comment]
Well that's a bummer. `WinUI 3` certainly looks like the `WPF` savior we all have been waiting for. I'm honestly very excited, but... creating a modern **Desktop app without docking panes**?! You have got to be kidding me!

This means we again have to go through the hassle of:

- Bloating our app with 10 MB of sloppy third-party code (they _never_ get the docking stuff right).
- Paying some license for a control which should have been there in the first place.
- Maintaining some abandoned DockManager toolkit.
- Writing our own docking system. Sure, but that means less time for cool features on our Windows app, and in the end less profit for Microsoft.

Name one other UI platform where you immediately need to look for commercial alternatives! iOS? Android? 😆

_"Now introducing WinUI 4 for Android. Scroll viewer? Nah, it's on our todo-list."_

**Comment by RetroHQ** [comment]
Just to keep the ball rolling on the fact that this is an absolutely essential feature.

I've been keeping an eye on the whole WinUI3 / Windows App SDK for a while and in particular docking windows / mdi. This is an absolutely essential feature for desktop productivity apps and any tool I have been considering writing of late. Quite literally every program I use for development on a daily basis uses this interface paradigm and it is absolutely essential.

Have I mentioned it's absolutely essential? I think I did.

For me, quite honestly, until there is SOME kind of docking / mdi type support (even 3rd party) it really does just feel like a toy for creating mobile-phone-wannabe-apps. 

MFC, some 30 years on, is still more attractive for those wanting to develop C++ based productivity apps on Windows. I'm not interested in moving to C# just for the UI only to jump through all the hoops required to then leverage the C++ codebase I require under the hood.

WinUI3 is nearly the thing of my dreams, finally bringing modern Windows GUI support back to C++. Make it happen. :)

**Comment by SkinnySackboy** [comment]
Commenting to keep this alive - this is absolutely essential for us to consider moving off WPF.

**Comment by sukesh-ak** [comment]
Any update on this?

I have a new project for which this is required. 
Played around with [UWP control from SyncFusion](https://www.syncfusion.com/uwp-ui-controls/docking). Time to jump on WinUI3 but this control is essential for the project.

**Comment by MaximMikhisor** [comment]
+1

**Comment by carmangary** [comment]
I definitely want to update this feature request.  Docking panels is an integral part of my MFC desktop application that I want to convert to WinUI3.

**Comment by MaximMikhisor** [comment]
> I definitely want to upvote this feature request. Docking panels is an integral part of my MFC desktop application that I want to convert to WinUI3.

Same things. We also want to convert WPF to WinUI3.

**Comment by fl0aty** [comment]
Please implement this feature. WinUI3 looks very promising, as UWP looked back in the day. But both are lacking the docking panes / MDF feature which is a bummer and is the sole reason to not move over from WPF.

A true, modern desktop application not only has to look good but also needs a lot of flexibility for the user to justify the need of a desktop application for him in the first place. There are a lot of business applications out there that need this kind of flexibility.

**Comment by SkinnySackboy** [comment]
@fl0aty for enterprise apps I cannot upvote your comment enough. It's basically a showstopper not having this for us.

**Comment by RetroHQ** [comment]
ImGui is a completely different UI paradigm, though, it's not always appropriate. The lack of support for this is totally bizarre to me. The only real option currently is having a C# front end with C++ back end, or using QT.

**Comment by RetroHQ** [comment]
Another random revisit of this topic as I consider using what UI API might be useful again in C++ land. And nothing appears to have changed on this absolutely essential feature. 🙄

It might be time to bite the bullet and use Qt… or just dig out the MFC libraries again.

**Comment by untimated** [comment]
came in 2023, seeing posts over years and still no progress really scares me to full dive to WinUi. Hope this feature will be out one day...

Having Official solution is always my preference over opensource or 3rd party solution. For now perhaps going back to win32 is the answer.

**Comment by carmangary** [comment]
I am puzzled by the lack of evolution of the Windows UI development stack.  WPF and WinUI seemed promising but more difficult to use than WinForms and MFC and did not have some of the critical features needed for rich business style applications.  I kind of lost interest in Windows development for that reason.  I'd love to see some sort of roadmap for Windows UI enrichment.

**Comment by kmilch** [comment]
Ditto carmangary!  What the heck?  Maybe I will move to QT.  Can QT host a Win32 window?

**Comment by jonasnordlund** [comment]
I honestly think the WinUI 3 tab control should be generalized to this. AvalonDock supports simply having them all fill their container as layered documents and then it basically becomes a tab control with tabs to change views.

The obvious benefit of this is that all WinUI apps would automatically gain docking support, although this could of course be opt-in on the control e.g. IsDockable.

**Comment by kokashking** [comment]
Its gone already 5 years since the proposal... dont feel like any cares....

**Comment by carmangary** [comment]
Seems as though Microsoft stopped caring about desktop developers, at least.  Azure is their cash cow now.

**Comment by znakeeye** [comment]
Now, three years later, I have to say that I no longer care about `WinUI` dock managers 😝.

We spent a few weeks trying to migrate a tiny version of our `WPF` application to `WinUI`. We're not convinced, to say the least.

Maybe in another three years...

**Comment by jonasnordlund** [comment]
Starting with .NET 9, WPF will have built-in support for good Windows 11 styling and controls, so a more sensible way forward here for complex LOB applications etc. might be WPF with [AvalonDock ](https://github.com/Dirkster99/AvalonDock) or a similar mature docking system. You can get decent Windows 11 ribbons via [FluentRibbon](https://github.com/fluentribbon/Fluent.Ribbon) 11.0 onwards, or [Coho.UI](https://github.com/sebbouez/Coho.UI).

**Comment by RetroHQ** [comment]
The whole point for me is that WinUI naively supports C++. If I wanted to jump through hoops and use C#, WPF, etc, then that's always been there. If you have a codebase you want to leverage in C++, it's far nicer to have the option to keep everything in C++. MFC or QT are really the only sensible solutions to do this. It really is insane, imo, that this has been left to stagnate as it has. I guess using C++ is not something the masses do these days?

**Comment by carmangary** [comment]
I agree with the above.  I have and still support a multimedia application written in C++ using MFC that has docking tabs.  I spent a lot of time porting it from C# due to C# being a pain to deal DirectX and also because it was so easily hackable due to ease of decompilation, even when obfuscated.  There was a noteable increase in application performance when moving to C++.

C# is cute and capable for a lot of stuff but it is not a good option for the type of development I want to do.  I'm not asking for much.  Just have WinUI match the level of capabilities that MFC had decades ago and improve upon it.

**Comment by RetroHQ** [comment]
This hits the point home for me. Can we just have the same level of features we had 20 years ago, please?

**Comment by kmilch** [comment]
VS2022 version 17.10.4 - Tried to create Blank App, C++ WinUI 3 in Desktop.  VS Created 911 files in 356 folders.  Tried to build and had 8 errors.  Good grief.

**Comment by Tartarusome** [comment]
> 从 .NET 9 开始，WPF 将内置对良好 Windows 11 样式和控件的支持，因此对于复杂的 LOB 应用程序等，更明智的前进方式可能是使用 [WPF 与 AvalonDock ](https://github.com/Dirkster99/AvalonDock) 或类似的成熟停靠系统。您可以通过 [FluentRibbon](https://github.com/fluentribbon/Fluent.Ribbon) 11 及更高版本（在我撰写本文时处于 alpha 阶段）或 [Coho.UI](https://github.com/sebbouez/Coho.UI) 获得不错的 Windows 11.0 功能区。如果您依赖早于 .NET 9 的 .NET 版本，您可以将 [WPFUI](https://github.com/lepoco/wpfui) 用于 Windows 11 样式，它支持低至 .NET Framework 4.6.2，实际上是 .NET 9 采用和集成的库。

I don't think so. Because Wpf has a lot of history problems.Such as using too many reflections, therefore is not able to support AOT. Meanwhile wpf is still a DiectX 9 thing, when Winui3 is based on DirectX 12, and Avalonia 11+ or C++ QT is based on Vulkan.
Just open a solution in vs2022's lastest version, which is a wpf app, and close this solution. Congratulations, you find a 100% memory leak.

**Comment by qian-o** [comment]
I hope it's not too late in 2025. Over the past few days, I have created a new project from scratch to implement the docking functionality. There has been some progress, and I hope everyone can provide valuable feedback to help me improve this project.

This project draws inspiration from several interesting open-source projects, such as AvalonDock and ImGui. However, the current code is still quite messy and needs to be organized. The first version is expected to be released in early March.

Project URL: https://github.com/qian-o/WinUI.Dock

https://github.com/user-attachments/assets/143a143c-b85e-4a3f-8d18-f675e643a7a5

**Comment by RetroHQ** [comment]
That  does look extremely promising! It's never too late.

**Comment by qian-o** [comment]
Hi, the WinUI.Dock control library I created has been released on [NuGet](https://www.nuget.org/packages/WinUI.Dock). You are welcome to try it out. It currently supports WinUI 3.0 and Uno Platform (partial functionality).

Currently, the control supports only a few features, but basic functionalities such as dragging, splitting, floating, and hiding have been implemented. Further improvements will be made, and suggestions and feedback are welcome.

https://github.com/user-attachments/assets/c73838d9-5a08-4ee4-a5ea-7aedfec12f81
